networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/BlockTopicHandler.java

```diff
@@ -62,14 +66,23 @@ protected SignedBeaconBlock deserialize(final Bytes bytes) throws SSZException {
 
   @Override
   protected boolean validateData(final SignedBeaconBlock block) {
+    if (chainStorageClient.isPreGenesis()) {
+      // We can't process blocks pre-genesis
+      return false;
+    }
+
     final BeaconState preState =
         chainStorageClient.getStore().getBlockState(block.getMessage().getParent_root());
     if (preState == null) {
-      // TODO - handle future and unattached blocks
-      LOG.warn(
-          "Dropping block message at slot {} with unknown parent state {}",
-          block.getMessage().getSlot(),
-          block.getMessage().getParent_root());
+      // Post event even if we don't have the prestate
+      eventBus.post(createEvent(block));
+      return false;
+    }
+
+    final UnsignedLong currentSlot = get_current_slot(chainStorageClient.getStore());
+    if (block.getSlot().compareTo(currentSlot) > 0) {
+      // Don't gossip future blocks
+      eventBus.post(createEvent(block));
```

Comment:
It feels very weird to say the data is invalid but process it anyway.  Kind of have to with the way gossip works but still awkward. Just something to ponder in the back of our minds about how it could be done better.

Reply:
yeah - it's not great.  i was thinking of reworking the API to surface some kind of `GossipMessage` object with some API methods like `doNotRegossip`, `regossip` or something but I felt like it wasn't worth digging into right now.